Chapter_3_Keyword_Design_Variables_Resources.md

```diff
@@ -0,0 +1,1027 @@
+
+
+#### 3.4.1.2 Named Test|Task With Multiple Data Rows:
+
+> [!IMPORTANT]
+> LO-XX Recall the syntax and properties of named test|task with multiple data rows (K1)
+
+A slightly different approach is to define multiple data rows for a single test|task.
+
+This is still possible with a single template defined in the `*** Settings ***` section, but in this case it would also make sense to define the template locally for each test|task with the `[Template]` setting.
+With this approach, it is possible to define different scenarios in the same suite file, which can be useful for testing different aspects of the same functionality.
+
+```robotframework
+*** Test Cases ***
+Invalid Logins
+    [Template]    Login With Invalid Credentials Should Fail
+    invalid          ${VALID PASSWORD}
+    ${VALID USER}    invalid
+    invalid          whatever
+    ${EMPTY}         ${VALID PASSWORD}
+    ${VALID USER}    ${EMPTY}
+    ${EMPTY}         ${EMPTY}
+
+Valid Logins
+    [Template]    Login With Valid Credentials Should Pass
+    ${VALID USER}            ${VALID PASSWORD}
+    ${VALID LONG USER}       ${VALID LONG PASSWORD}
+    ${VALID COMPLEX USER}    ${VALID COMPLEX PASSWORD}
+```
+
+If one data row fails, this template execution is marked FAIL and the test|task is marked FAIL, but **the other data rows are still executed**.
+
+This approach creates only a single tests|tasks for multiple data rows in the logs and reports, which can be beneficial statistically.
+
+However, this approach has also its drawbacks:
+
+- Test|task setup and teardown are executed only once for all data rows of one test|task.
+  If there is a setup and teardown needed for each data row, a keyword setup or teardown is needed.
+- The test|task name is not unique for each data row, which can make it harder to understand the failing data row in the logs.
+- Filtering and re-execution of some or single data rows is not possible.
+
+
+
+
+
+## 3.5 Advanced Importing of Keywords and Naming Conflicts
+
+> [!IMPORTANT]
+> LO-XX Recall that naming conflicts can arise from the import of multiple resource files (K1).
+
+As stated before, it is possible to organize imports and available keywords in Robot Framework by using Resource Files.
+By default, all keywords or variables created or imported in a resource file are available to those suites and files that are importing that higher-level resource file.
+
+This can lead to complex import hierarchies or the importing of libraries multiple times, which should be avoided.
+
+Due to this mechanism, the number of keywords available to a suite can be quite large, and naming conflicts, especially with keywords from third-party keyword libraries, can occur. These conflicts need to be resolved.
+
+
+Some keyword libraries have the option to be configured to change their behavior, which may also change the available keywords they offer.
+
+
+
+### 3.5.1 Importing Hierarchies
```

Comment:
I do not see the recursion in this example...?

=> I also think that this section should rather show what happens when libraries are imported multiple times. Recursion is not really necessary to show this. This can also happen when Library A gets imported from both resource files.